server/src/main/java/org/candlepin/controller/CandlepinPoolManager.java

```diff
@@ -1597,105 +1624,30 @@ public Entitlement adjustEntitlementQuantity(Consumer consumer, Entitlement enti
         Map<String, Integer> poolQuantityMap, CallerType caller)
         throws EntitlementRefusedException {
 
-        // Because there are several paths to this one place where entitlements
-        // are granted, we cannot be positive the caller obtained a lock on the
-        // pool when it was read. As such we're going to reload it with a lock
-        // before starting this process.
-        log.debug("Locking pools: {}", poolQuantityMap.keySet());
-        Collection<Pool> pools = poolCurator.lockAndLoadByIds(poolQuantityMap.keySet());
-
-        if (log.isDebugEnabled()) {
-            for (Pool pool : pools) {
-                log.debug("Locked pool: {} consumed: {}", pool, pool.getConsumed());
-            }
-        }
-
-        log.debug("Done locking pools");
+        BindChain chain = new BindChain();
+        BindContext context = bindContextFactory.create(consumer, poolQuantityMap);
+        chain.addOperation(rulesCheckOpFactory.create(caller));
+        chain.addOperation(handleEntitlementsOp);
+        chain.addOperation(postBindBonusPoolsOp);
+        chain.addOperation(handleCertificatesOp);
+        chain.addOperation(complianceOp);
 
```

Comment:
@mstead  looking at what I ended up with, I just added a commit to this PR. 
@awood  and @mstead  please review the latest commit .

@mstead  ,  I actually spent the last hour trying to write a generic chain before I saw your comment, but I think that is not a good idea right now, this is what I found:

We can create a generic Operation interface, and OpChain abstract class in an "operations" package, and then implement them as BindChain, while each operation could extend Operation, but the problem lies with BindContext.
both preprocess and execute in the bind chain need the actual BindContext, and not a generic context.
This problem can be solved, but I think we should wait till we have atleast one different type of chain. only then we can pick one of the two following approaches to solve the problem:
1. change BindContext to hold all entities in a map so it can be shared among all kinds of chains.
2. based on what entities the other chain would need,  we could perhaps use the BindContext as is.

I am hesitant to just say that no matter what chains we have, we will use a generic context, cause then we lose encapsulation imo. so I thought it best to simply defer this for future when things will get clearer once we know what our new chain will look like.